cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/GitHubHttpHandler.java

```diff
@@ -0,0 +1,453 @@
+  private static final Gson GSON = new Gson();
+  private final GitHubStore gitStore;
+
+  private final SecureStore secureStore;
+  private final SecureStoreManager secureStoreManager;
+
+
+  @Inject
+  GitHubHttpHandler(GitHubStore gitStore, SecureStore secureStore,
+      SecureStoreManager secureStoreManager) {
+    this.gitStore = gitStore;
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+  }
+
+  @Path("repos/github")
+  @GET
+  public void getRepos(HttpRequest request, HttpResponder responder) throws Exception {
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gitStore.getRepos()));
+  }
+
+  /**
+   * Returns user repository information
+   */
+  @Path("repos/github/{repo}")
+  @GET
+  public void getRepoInfo(HttpRequest request, HttpResponder responder,
+                          @NotNull @PathParam("repo") String repo) throws Exception {
+    if (gitStore.getRepo(repo) != null) {
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gitStore.getRepo(repo)));
+    } else {
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+    }
+  }
+
+  @Path("repos/github/{repo}")
+  @PUT
+  public void addRepoInfo(FullHttpRequest request, HttpResponder responder,
+                          @NotNull @PathParam("repo") String repo) throws Exception {
+
+      GitHubRepo githubRequest = GSON.fromJson(request.content().toString(StandardCharsets.UTF_8)
+          , GitHubRepo.class);
+
+      if (githubRequest.validateAllFields()) {
+        //test the repo's connection to ensure it exists.
+        int responseCode = testRepoConnection(githubRequest,
+            "token " + githubRequest.getAuthString());
+        if (responseCode == HttpURLConnection.HTTP_OK) {
+          String authString = "token " + githubRequest.getAuthString();
+          SecureKeyId authKeyId = new SecureKeyId("system", repo + "_auth_token");
+          secureStoreManager.put("system", repo + "_auth_token", authString,
+              repo + "Authorization Token", new HashMap<String, String>());
+          gitStore.addOrUpdateRepo(repo, githubRequest.getUrl(), githubRequest.getDefaultBranch(),
+              repo + "_auth_token");
+          responder.sendString(HttpResponseStatus.OK, "Repository Information Saved.");
+        } else if (responseCode == HttpURLConnection.HTTP_MOVED_PERM) {
+          responder.sendString(HttpResponseStatus.MOVED_PERMANENTLY, "Repository has been moved.");
+        } else if (responseCode == HttpURLConnection.HTTP_FORBIDDEN) {
+          responder.sendString(HttpResponseStatus.FORBIDDEN, "You do not have access to this repository.");
+        } else {
+          responder.sendString(HttpResponseStatus.NOT_FOUND, "Repository was not found.");
+        }
+        //If the user leaves fields blank, throw back which fields are missing
+      } else {
+        String errorString = "Please enter a ";
+        if (!githubRequest.validNickname()) {
```

Comment:
so here are you doing the validation again , since you already called `validateAllFields()`.

You have two ways  to do this :

1. validate each fields one by one and shortcut if any of the field is invalid like:
   List<String> errorFields = new ArrayList<>();    
if (! validateXXX() ) {
        errorField.add(...)
   }
   if  (! validate XXX() ) {
      errorField.add(...)
   }
    ....
  if ( !errorField.isEmpty()) {
        respond with error
       return; 
  }
  normal logic goes here 

2. use an error collector to collect errors. Ref : https://github.com/cdapio/cdap/blob/develop/cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/FailureCollector.java
  still call validateAllFields() and pass in the failure collector to collect all errors

Reply:
the latter one is better. so that your code here does not need to be aware of what field is being validated.

Reply:
I tried doing the latter, but for some reason could not import FailureCollector into the handler, so I implemented option 1.